File: app/src/main/java/com/nbennettsoftware/android/npad/storage/WallpaperManager.java
package com.nbennettsoftware.android.npad.storage;

import android.content.ContentResolver;
import android.content.Context;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.provider.MediaStore;

import com.nbennettsoftware.android.npad.R;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class WallpaperManager {
    private Context context;
    private String prefs_key_internal_wallpaper_name;

    public WallpaperManager(Context context) {
        this.context = context;
        prefs_key_internal_wallpaper_name = context.getResources().getString(R.string.pref_key_internal_wallpaper_name);
    }

    public void replaceInternalizeWallpaper(Uri wallpaperUri) throws ReplaceInternalWallpaperException {
        try {
            ContentResolver resolver = context.getContentResolver();
            String displayName = getWallpaperFileName(wallpaperUri);

            deleteInternalizedWallpaper();

            //Copy the wallpaper to internal storage
            OutputStream outputStream = context.openFileOutput(displayName,0);
            InputStream inputStream = resolver.openInputStream(wallpaperUri);

            //Once an number of bytes are read, write said bytes to output.
            byte[] buffer = new byte[1024];
            int bytesRead;
            while ((bytesRead = inputStream.read(buffer)) != -1 ){
                outputStream.write(buffer, 0, bytesRead);
            }

            outputStream.close();
            inputStream.close();

            //Store new wallpaper name in prefs.
            getPreferences().edit()
                    .putString(prefs_key_internal_wallpaper_name, displayName)
                    .apply();

        } catch (IOException | UriDataRetrievalException e ) {
            throw new ReplaceInternalWallpaperException();
        }
    }

    public class ReplaceInternalWallpaperException extends Exception {}

    public void deleteInternalizedWallpaper() {
        try {
            File internalWallpaper = getInternalizedWallpaper();
            internalWallpaper.delete();
            getPreferences().edit().remove(prefs_key_internal_wallpaper_name).apply();
        } catch (NoInternalWallpaperException e) {
            //Do nothing
        }
    }

    public class NoInternalWallpaperException extends Exception{};

    public File getInternalizedWallpaper() throws NoInternalWallpaperException {
        String fileName = getPreferences().getString(prefs_key_internal_wallpaper_name, null);
        if(fileName==null){ throw new NoInternalWallpaperException(); }
        String internalDirPath = context.getFilesDir().getAbsolutePath();
        return new File(internalDirPath, fileName);
    }

    private SharedPreferences getPreferences(){
        return PreferenceManager.getDefaultSharedPreferences(context);
    }

    private class UriDataRetrievalException extends Exception{}

    private String getWallpaperFileName(Uri uri) throws UriDataRetrievalException {
        String stringData=null;
        ContentResolver resolver = context.getContentResolver();
        String[] projection = {MediaStore.MediaColumns.DISPLAY_NAME};
        Cursor cursor = resolver.query(uri,projection,null,null,null);
        if(cursor==null) { throw new UriDataRetrievalException(); }
        if(cursor.moveToFirst()){
            int columnId = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DISPLAY_NAME);
            stringData = cursor.getString(columnId);
        }
        cursor.close();
        return stringData;
    }

}

File: app/src/main/java/com/nbennettsoftware/android/npad/storage/WallpaperManager.kt
package com.nbennettsoftware.android.npad.storage

import android.content.Context
import android.content.SharedPreferences
import androidx.preference.PreferenceManager
import android.net.Uri
import android.provider.MediaStore

import com.nbennettsoftware.android.npad.R

import java.io.File
import java.io.IOException

class WallpaperManager(private val context: Context) {

    private val prefsKeyInternalWallpaperName: String = context.resources.getString(R.string.pref_key_internal_wallpaper_name)

    val internalizedWallpaper: File
        @Throws(NoInternalWallpaperException::class)
        get() {
            val fileName = preferences.getString(prefsKeyInternalWallpaperName, null)
                    ?: throw NoInternalWallpaperException()
            val internalDirPath = context.filesDir.absolutePath
            return File(internalDirPath, fileName)
        }

    private val preferences: SharedPreferences
        get() = PreferenceManager.getDefaultSharedPreferences(context)

    @Throws(ReplaceInternalWallpaperException::class)
    fun replaceInternalizeWallpaper(wallpaperUri: Uri) {
        try {
            deleteInternalizedWallpaper()

            val displayName = getWallpaperFileName(wallpaperUri)
            val contentResolver = context.contentResolver

            //Copy the wallpaper to internal storage
            val outputStream = context.openFileOutput(displayName, 0)
            val inputStream = contentResolver.openInputStream(wallpaperUri)

            inputStream.use { input ->
                outputStream.use { output ->
                    input!!.copyTo(output)
                }
            }

            outputStream.close()
            inputStream!!.close()

            //Store new wallpaper name in prefs.
            preferences.edit()
                    .putString(prefsKeyInternalWallpaperName, displayName)
                    .apply()
        } catch (e: IOException) {
            throw ReplaceInternalWallpaperException()
        } catch (e: UriDataRetrievalException) {
            throw ReplaceInternalWallpaperException()
        }

    }

    inner class ReplaceInternalWallpaperException : Exception()

    fun deleteInternalizedWallpaper() {
        try {
            val internalWallpaper = internalizedWallpaper
            internalWallpaper.delete()
            preferences.edit().remove(prefsKeyInternalWallpaperName).apply()
        } catch (e: NoInternalWallpaperException) {
            //Do nothing
        }

    }

    fun hasInternalizedWallpaper(): Boolean {
        val preferences = preferences
        val wallpaper = preferences.getString(prefsKeyInternalWallpaperName, null)
        return wallpaper != null && wallpaper.isNotEmpty()
    }

    inner class NoInternalWallpaperException : Exception()

    private inner class UriDataRetrievalException : Exception()

    @Throws(UriDataRetrievalException::class)
    private fun getWallpaperFileName(uri: Uri): String? {

        lateinit var stringData: String
        val contentResolver = context.contentResolver
        val projection = arrayOf(MediaStore.MediaColumns.DISPLAY_NAME)

        contentResolver.query(uri, projection, null, null, null)?.use { cursor ->
            if (cursor.moveToFirst()) {
                val columnId = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DISPLAY_NAME)
                stringData = cursor.getString(columnId)
            }
        }?: throw UriDataRetrievalException()

        return stringData
    }
}
